src/ddmd/expressionsem.d

```diff
@@ -7859,16 +7859,47 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             }
         }
 
+        Type t1 = exp.e1.type.toBasetype();
+        Type t2 = exp.e2.type.toBasetype();
+
+
+        bool needsDirectEq()
```

Comment:
could you refactor this to take parameters and not depend on internal state ?
such a refactoring would increase the reusablity.